espnet2/asr/frontend/espnet_model.py

```diff
@@ -0,0 +1,416 @@
+                )
+            else:
+                mask_pre_ = [
+                    mask_pre["spk{}".format(spk + 1)] for spk in range(self.num_spk)
+                ]
+
+                # compute TF masking loss
+                # TODO:Chenda, Shall we add options for computing loss on the masked spectrum?
+                tf_loss, perm = self._permutation_loss(
+                    mask_ref, mask_pre_, self.tf_mse_loss
+                )
+
+                if "dereverb" in mask_pre:
+                    if dereverb_speech_ref is None:
+                        raise ValueError(
+                            "No dereverberated reference for training!\n"
+                            'Please specify "--use_dereverb_ref true" in run.sh'
+                        )
+                    tf_loss = (
+                        tf_loss
+                        + self.tf_l1_loss(
+                            dereverb_mask_ref, mask_pre["dereverb"]
+                        ).mean()
+                    )
+
+                if "noise1" in mask_pre:
+                    if noise_ref is None:
+                        raise ValueError(
+                            "No noise reference for training!\n"
+                            'Please specify "--use_noise_ref true" in run.sh'
+                        )
+                    mask_noise_pre = [
+                        mask_pre["noise{}".format(n + 1)]
+                        for n in range(self.num_noise_type)
+                    ]
+                    tf_noise_loss, perm_n = self._permutation_loss(
+                        noise_mask_ref, mask_noise_pre, self.tf_mse_loss
+                    )
+                    tf_loss = tf_loss + tf_noise_loss
+
+            if self.training and self.tf_factor == 1.0:
+                si_snr_loss = None
+                si_snr = None
+            else:
+                speech_pre = [
+                    self.frontend.stft.inverse(ps, speech_lengths)[0]
+                    for ps in spectrum_pre
+                ]
+                if speech_ref[0].dim() == 3:
+                    # For si_snr loss, only select one channel as the reference
+                    speech_ref = [sr[..., self.ref_channel] for sr in speech_ref]
+                # compute si-snr loss
+                si_snr_loss, perm = self._permutation_loss(
+                    speech_ref, speech_pre, self.si_snr_loss, perm=perm
+                )
+                si_snr = -si_snr_loss
+
+            if self.tf_factor != 1.0:
+                loss = (1 - self.tf_factor) * si_snr_loss + self.tf_factor * tf_loss
+            else:
+                loss = tf_loss
+
+            stats = dict(
+                si_snr=si_snr.detach() if si_snr is not None else None,
+                tf_loss=tf_loss.detach(),
+                loss=loss.detach(),
+            )
+        else:
+            # TODO:Jing, should find better way to configure for the choice of tf loss and time-only loss.
```

Comment:
This is just a note.
A possible extension is to combine both losses as a multitask fashion, but if it makes the code very complicated, we don't have to consider it.